### PR TITLE
Fix OpenAPIValueContainer serialization of nested values

### PR DESCRIPTION
**Pull request by czechboy0**
### Motivation

Credit goes to @rboyce for discovering this bug!

When I saw the PR #24, I realized we didn't have any unit tests for the `OpenAPI*Container` types, which was an oversight.

This PR adds those missing unit tests, and fixes the bug in casting that motivated #24.

### Modifications

Adds unit tests and fixes the casting bug that expected wrapped, instead of raw stdlib values.

### Result

Now all the new unit tests pass and serializing nested values in OpenAPIValueContainer and friends works correctly.

### Test Plan

Most of this PR is new unit tests, verified they pass locally.

### CODE REVIEW

**Comment by czechboy0** [comment]
@rboyce once this PR lands, please rebase your PR and add the great unit tests you had there to the basic ones I added in this PR. 🙏

**Comment by czechboy0** [comment]
@simonjbeaumont let's get #22 landed first, then I'll update this PR, and then a subset of #24 with just the additional unit tests.

File: Sources/OpenAPIRuntime/Base/OpenAPIValue.swift [hunk 1]
```diff
@@ -139,9 +139,9 @@ public struct OpenAPIValueContainer: Codable, Equatable, Hashable, Sendable {
             try container.encode(value)
         case let value as String:
             try container.encode(value)
-        case let value as [OpenAPIValueContainer?]:
+        case let value as [Sendable?]:
             try container.encode(value.map(OpenAPIValueContainer.init(validatedValue:)))
-        case let value as [String: OpenAPIValueContainer?]:
+        case let value as [String: Sendable?]:
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Base/OpenAPIValue.swift:144]:
```suggestion
        case let value as [String: (any Sendable)?]:
```

And elsewhere

Comment (czechboy0) [reply Sources/OpenAPIRuntime/Base/OpenAPIValue.swift:144]: Ah good reminder, I opened #26 to land that first, then I'll update this PR.

**Comment by czechboy0** [comment]
Investigating the integration test failure, we're seeing behavior that I don't know whether it's desired or a bug in SwiftPM.

When we enable the existential any feature in the runtime library, it also forces that feature on in the integration test package, which depends on the runtime library. That'd mean that enabling an upcoming feature in an upstream package forces it on all downstream packages?

Will try to create a reproduction case to see if we're just doing something wrong. But tagging @neonichu in case this is known.

**Comment by neonichu** [comment]
@DougGregor may know more here

**Comment by DougGregor** [comment]
> @DougGregor may know more here

That's not supposed to be the case. Upcoming features should only affect the target, not downstream targets.

**Comment by czechboy0** [comment]
@neonichu @DougGregor I spent a few hours bisecting this using toolchains from Swift.org, and it seems to have been fixed between 04-26 and 04-27. Before it, including 5.8.1, this bug is occurring. After, it's not. Unfortunately, 5.8.1 will need to be supported for a few more months in this project, so I wonder if there's a chance of a cherry pick + 5.8.2?

swift-package-manager and swift-driver don't have any changes that look related to me, but the main Swift repo has stuff related to existential any, which is the feature we're seeing "leaked" from an upstream to downstream dependees.

https://github.com/apple/swift/compare/swift-5.9-DEVELOPMENT-SNAPSHOT-2023-04-26-a...swift-5.9-DEVELOPMENT-SNAPSHOT-2023-04-27-a

**Comment by czechboy0** [comment]
To recap - what we're seeing on these older Swift versions is that when we enable the ExistentialAny feature in the upstream Runtime package, it affects a downstream integration test package, which consumes the Runtime library.

When we disable it in Runtime, it goes back to working.

**Comment by czechboy0** [comment]
API breakage is the known sendable stuff

```
2 breaking changes detected in OpenAPIRuntime:
  💔 API breakage: constructor OpenAPIObjectContainer.init(unvalidatedValue:) has parameter 0 type change from [Swift.String : Any?] to [Swift.String : (Swift.Sendable)?]
  💔 API breakage: constructor OpenAPIArrayContainer.init(unvalidatedValue:) has parameter 0 type change from [Any?] to [(Swift.Sendable)?]
** ERROR: ❌ Breaking API changes detected.
```

**Review by simonjbeaumont — APPROVED** [review]

**Pull request merged**